### PR TITLE
feat: use separate nonces

### PR DESCRIPTION
**Pull request by Jean-Grimal**

### CODE REVIEW

File: src/ERC20DelegatesUpgradeable.sol [hunk 1]
```diff
@@ -55,6 +49,11 @@ abstract contract ERC20DelegatesUpgradeable is
         return $._delegatee[account];
     }
 
+    function delegationNonces(address account) public view returns (uint256) {
```
Comment (adhusson) [inline src/ERC20DelegatesUpgradeable.sol:52]:
```suggestion
    function delegationNonces(address account) external view returns (uint256) {
```

(should be moved to `/* EXTERNAL */`)

File: test/MorphoToken.t.sol [hunk 2]
```diff
@@ -141,7 +141,8 @@ contract MorphoTokenTest is BaseTest {
         assertEq(newMorpho.delegates(delegator), delegation.delegatee);
         assertEq(newMorpho.getVotes(delegator), 0);
         assertEq(newMorpho.getVotes(delegation.delegatee), amount);
-        assertEq(newMorpho.nonces(delegator), 1);
+        assertEq(newMorpho.delegationNonces(delegator), 1);
```
Comment (adhusson) [inline test/MorphoToken.t.sol:144]: Missing a test that increases nonces but not delegationNonces.

File: src/interfaces/IDelegates.sol [hunk 3]
```diff
@@ -9,6 +9,9 @@ interface IDelegates {
     // @dev The signature used has expired.
     error DelegatesExpiredSignature(uint256 expiry);
 
+    // @dev The nonce used for an `account` is not the expected current nonce.
```
Comment (adhusson) [inline src/interfaces/IDelegates.sol:12]:
```suggestion
    // @dev The delegation nonce used by `account` is not its current delegation nonce.
```

**Review by adhusson — CHANGES_REQUESTED** [review]

**Review by adhusson — APPROVED** [review]

File: src/ERC20DelegatesUpgradeable.sol [hunk 4]
```diff
@@ -141,6 +140,15 @@ abstract contract ERC20DelegatesUpgradeable is
         }
     }
 
+    function _useCheckedDelegationNonce(address owner, uint256 nonce) private {
+        ERC20DelegatesStorage storage $ = _getERC20DelegatesStorage();
+        uint256 current = $._nonces[owner];
+        if (nonce != current) {
+            revert InvalidDelegationNonce(owner, current);
+        }
```
[the remark this answers is not in the document]
Comment (MathisGD) [reply src/ERC20DelegatesUpgradeable.sol:148]:
```suggestion
        require(nonce == $._nonces[owner], InvalidDelegationNonce(owner, current));
```

File: src/ERC20DelegatesUpgradeable.sol [hunk 5]
```diff
@@ -141,6 +140,15 @@ abstract contract ERC20DelegatesUpgradeable is
         }
     }
 
+    function _useCheckedDelegationNonce(address owner, uint256 nonce) private {
```
Comment (MathisGD) [inline src/ERC20DelegatesUpgradeable.sol:143]: could we inline it?

**Review by MathisGD — APPROVED** [review]

**Review by MerlinEgalite — COMMENTED** [review]
LGTM

**Pull request merged**